### ISSUE TITLE
Unicode Character Map - 0x02FF (a.k.a. "ː") Details

### ISSUE THREAD

**Issue by h4ck3rm1k3** [opal/opal#1505]
I am getting triangular : in the generated javascript 
Unicode Hexadecimal: 0x02D0 
Unicode Decimal: 720 
UCS-2 Hexadecimal: 0xD002 
UCS-2 Decimal: 53250 
HTML Hexadecimal: &#x02D0; 
HTML Decimal: &#720; 
http://www.unicodemap.org/details/0x02D0/index.html

example 
```
from opal/opal/corelib/helpers.rb:1:in source
....
    Opal.defs(self, '$type_error', TMP_2 = function ːtype_error(object, type, method, coerced) {
```
So the name itself is modified.

The original is 
```
 def self.type_error(object, type, method = nil, coerced = nil)
```

It occurs on all class names.

Looking at the file in the browser, it displays as 
```
 Opal.defs(self, '$try_convert', TMP_6 = function Ëtry_convert(object, type, method) {
```

I found the source here 
opal/lib/opal/nodes/def.rb:90:Opal::Nodes::DefsNode:<:         function_name = valid_name?(mid) ? " ː#{mid}" : ''

https://github.com/opal/opal/blob/6d5fe582991037e6a75ee4828dc4d285f7339c1d/lib/opal/nodes/def.rb#L86

It says this is a special utf char,

**Comment by iliabylich** [comment]
@elia How about using underscore as a function prefix?

**Comment by h4ck3rm1k3** [comment]
I changed it to $_ and it worked.

**Comment by iliabylich** [comment]
@h4ck3rm1k3 Yes, any combination of `$` and `_` fixes your issue. The question is: can we use just `_` and can we get a conflict with local variables in this case? With a special character `ː` we definitely can't, but it doesn't work in 100% of cases :smile:

**Comment by h4ck3rm1k3** [comment]
So I am wondering if opal is using the https://github.com/ruby/spec because when I tried to run that, it produced huge number of errors. It supposedly supports opal. I don't know if we have full coverage.

**Comment by iliabylich** [comment]
@h4ck3rm1k3 How did you run it?

**Comment by h4ck3rm1k3** [comment]
ruby  -I ~/experiments/opal/lib/  ~/experiments/opal/bin/opal

On Mon, Jul 4, 2016 at 8:08 AM, Ilya Bylich <notifications@github.com>
wrote:

> @h4ck3rm1k3 <https://github.com/h4ck3rm1k3> How did you run it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opal/opal/issues/1505#issuecomment-230276046>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AACIVzI-nPuMbo3JDzye2O3owjUO63iDks5qSPfYgaJpZM4JEQ22>
> .
>



-- 
James Michael DuPont
Kansas Linux Fest http://kansaslinuxfest.us
Free/Libre Open Source and Open Knowledge Association of Kansas
http://openkansas.us
Member of Free Libre Open Source Software Kosova http://www.flossk.org
Saving Wikipedia(tm) articles from deletion http://SpeedyDeletion.wikia.com

**Comment by iliabylich** [comment]
@h4ck3rm1k3 Could you try to follow this guide - https://github.com/opal/opal/blob/master/HACKING.md

**Comment by elia** [comment]
It's important to set the encoding in the html document loading opal code, not sure that's the issue, but sounds likely. See here for an example: https://github.com/opal/opal#compiling-ruby-code-with-the-cli-command-line-interface

**Comment by h4ck3rm1k3** [comment]
Ok, well I have not looked deeply into this, but one the cli I am also
getting the strange chars, I guess I need to set my locale. Is there any
reason why we need unicode for this?

On Mon, Jul 4, 2016 at 6:44 PM, Elia Schito <notifications@github.com>
wrote:

> It's important to set the encoding in the html document loading opal code,
> not sure that's the issue, but sounds likely. See here for an example:
> https://github.com/opal/opal#compiling-ruby-code-with-the-cli-command-line-interface
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opal/opal/issues/1505#issuecomment-230361066>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AACIV_mWife_7rKadYcxDDRiTUcdcZ6Lks5qSYzWgaJpZM4JEQ22>
> .
>



-- 
James Michael DuPont
Kansas Linux Fest http://kansaslinuxfest.us
Free/Libre Open Source and Open Knowledge Association of Kansas
http://openkansas.us
Member of Free Libre Open Source Software Kosova http://www.flossk.org
Saving Wikipedia(tm) articles from deletion http://SpeedyDeletion.wikia.com

**Comment by elia** [comment]
@h4ck3rm1k3 I think a double dollar `$$` could be used instead, but at the time of choosing which char to use I thought to be ok to require `utf8` as some stuff behaves weirdly otherwise (e.g. some string & regexp related methods) and that could help avoid difficult to debug errors down the road.

**Comment by elia** [comment]
Closing, this, `utf8` will be required for now, also on master now the special char is `ⵌ` (for similarity with `#`).

**Comment by svoboda-jan** [comment]
@elia using `UTF-8` characters in function names seems to break advertised compatibility with `Internet Explorer 6+`.

Just came across an error of `Invalid character` in IE 6 when loading opal.

I don't think it is worth introducing `UTF-8` dependency just to have fancy function names.

**Comment by h4ck3rm1k3** [comment]
I have had a bunch of my bug reports and prs closed on this project for no apparent reason, it is one reason why I stopped sending you them. Please consider if you want other contributors and treat them accordingly.

**Comment by svoboda-jan** [comment]
__IE 6 update__: It works when the `UTF-8` character is not at the beginning of the function name. Also other `UTF-8` characters (like `č`) seem to work even at the beginning of the function name.

But even if prefixing the `UTF-8` character would work for IE 6, I still don't think it is worth introducing `UTF-8` dependency just to have fancy function names.

**Comment by svoboda-jan** [comment]
@elia is there any list or documentation of methods that would "behave wrongly" ? I might be able to help with that. Also IE6 does support `UTF-8` encoding, the problem is using it in function names.

If the reason for `UTF-8` is not to have fancy function names, then I would suggest changing the character to something else like `_RBM_` or any `UTF-8` character that works. It would solve IE 6 compatibility and the `UTF-8` requirement may stay.

**Comment by elia** [comment]
@svoboda-jan working on browser testing I already rolled back that change as is was the only thing blocking IE6 from running (see https://github.com/opal/opal/tree/elia/browser-testing?files=1), namely I switched to a double dollar `$$`. 

About the methods I don't remember to be honest, I think it was 1 or 2 years ago.

**Comment by svoboda-jan** [comment]
@elia 👍 

About the string methods, I haven't came across anything yet and I am using languages that require it.

**Comment by zetachang** [comment]
Got the `Uncaught SyntaxError: Invalid or unexpected token` error. And it took me hours to finally realized that the browser (Chrome) couldn't auto detect encoding in setup like [the rack example](https://github.com/opal/opal/tree/master/examples/rack).

**Comment by elia** [comment]
I've pushed 9e32638d9b (Avoid special utf8 chars & be nice to older IEs) to opal/opal@0-10-stable, that when released as v0.10.3 should solve all these issues. The rationale is that the error you get is so bad that worse than going on with a page that has the wrong encoding. 🙂
